### DISCUSSION TITLE
Expose image filenames when asset presets are configured

### DISCUSSION THREAD

**Discussion by aileksandar** [directus/directus#9079] [Feature Requests]
I know there is a lot of discussion here about having direct access to storage urls, and this is maybe not the best title, but hear me out.

Directus way of exposing images and assets is by hitting the `/assets/:id` endpoint, and it gives you a lot of flexibility in terms of image transformation and permissions, and I really love that features. But, I currently use an B2 (Backblaze via S3 api) bucket for storage, and that approach is becoming a bottleneck for me.

1. My server is in US-West and my B2 bucket is in EU, so currently image loading is very slow because the server is pulling images from the bucket into memory and then streaming it to the browser.
2. Also, correct me if I'm wrong, the bandwidth cost is doubled with that approach, because server is spending your bucket bandwidth and client is spending your server bandwidth, and that can become a big difference in cost specially now that Cloudflare announced their R2 with 0 egress cost, or in my case B2 that is significantly cheaper than S3.

I know I don't have to use the `/assets` endpoint and I can just set up the frontend to load images from `<s3 url> + <filename_disk>`, but I really like that Directus is handling the thumbnails. Same problem that @rijkvanzanten mentioned here https://github.com/directus/directus/discussions/7612#discussioncomment-1241292

So the compromise IMO would be exposing the filenames for images that has been transformed with some asset preset. Let's say you made 2 presets: `standard` and `thumb`. I would add to file object some virtual attributes, for example `filename_disk_standard` and `filename_disk_thumb` and that way I could access the bucket URL directly and also keep some of the transform functionality. 

I know that you would not be able to on the fly transformation via query params but that the compromise that I'm talking about. Second compromise is that you would also not be able to have permissions once the link is out there, but to be honest, I'm not sure how needed is that for images. Even on Facebook you can copy the image URL and send it to someone as public. Permissions URL makes sense IMO for files, and for that it makes sense still using the `/assets` endpoint.

I tried to make some proof of concept as a hook, but I cannot update the response because the `items.read.before` hook is not enabled. I know that's because you [can kill the API ](https://discord.com/channels/725371605378924594/741327927051813108/857653313288536094) but IMO, you can kill the app with so many different things in hooks and custom endpoints, that responsibility should be on the developer too.

```js
const hash = require('object-hash');
const path = require('path');
const TransformationUtils = require('directus/dist/utils/transformations');

module.exports = function registerHook({ services }) {
  const { SettingsService } = services;
  return {
    'items.read': async function ({ payload, schema }) {
      const settings = new SettingsService({ schema });
      const savedSettings = await settings.readOne(1);

      payload = payload.map((item) => {
        item = savedSettings.storage_asset_presets.map(preset => {
          item.image[`url_${preset.key}`] = getAssetFilename(item.image, preset)
          return item;
        })
        return item;
      });
    },
  };
};

const getAssetFilename = (file, transformation) => {
  const transforms = TransformationUtils.resolvePreset(transformation, file);

  const maybeNewFormat = TransformationUtils.maybeExtractFormat(transforms);

  return (
    path.basename(file.filename_disk, path.extname(file.filename_disk)) +
    getAssetSuffix(transforms) +
    (maybeNewFormat ? `.${maybeNewFormat}` : path.extname(file.filename_disk))
  );
};

const getAssetSuffix = (transforms) => {
  if (Object.keys(transforms).length === 0) return '';
  return `__${hash(transforms)}`;
};
```

This is how I tried to expose the names by using the current logic in Directus. And I'm pretty sure it would be working if `item.read.before` was possible, but we can make another discussion for that.

I really think this is a problem worth solving, I would like to hear your thoughts.

Cheers

**Comment by rijkvanzanten** [comment]
Like it!

> I really think this is a problem worth solving, I would like to hear your thoughts.

I would agree. It sounds like we can take away the pain point in three steps:

1) Generate the (presets-based) thumbnails on upload instead of on request
2) Expose a direct link to the original file (ref https://github.com/directus/directus/pull/8065)
3) Add a "recognizable" name for each configured preset [^1]

Does that sound about right to you?

[^1]: Right now, we generate thumbnails with "filename" + a hash of the transformations object. In order to access the thumbnails properly, we should generate preset-thumbnails with a "guessable" name instead, so you could simply open "s3 base" + "filename" + "preset"

**Reply by aileksandar** [reply]
> Generate the (presets-based) thumbnails on upload instead of on request

Yes, this will solve another issue that I have. Currently, after the image is uploaded, the user is still waiting for the image to transform, but it doesn't see the loader because the upload API returned the response. So you might consider to first generate thumbs and then return the response, or to make it as a param that files API would receive.

> Expose a direct link to the original file (ref [Direct Asset URL to Storage #8065](https://github.com/directus/directus/pull/8065))

Not sure if this completely correlates with this idea, seems like its more focused on images in text editor, and it relies on database field, I'll take a look.

> Add a "recognizable" name for each configured preset

I would keep the filename on storage the same as it is now to keep the backward compatibility, just add "friendly" names as virtual fields in the file object, same as I did in the hook on the comment above.

I guess this is not really a priority for you, so I can try and spin up a PR. I'll first take a look at the asset URL PR and then start working on this when we agree on all the points.

**Comment by luciandavidescu** [comment]
For just about any static setting this couldn't feasibly be done without...
Simplest use-case: Spin a stack on your localhost, commit a blog post to _gh-pages_ and such, close the laptop and be on your way.
Middlish (mine*): Expect to at least have a fallback if the dynamic features somehow fail.
Complex: Set up multiple hosts or CDNs to load-balance in case of peak/ddos/downtime. Or choose to rely exclusively on a specialized webserver, for high-concurrency contexts.
*I noticed the issue when for some reason my _access_token_ got stripped and I got the broken image icon.

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
